### PR TITLE
HIVE-27589: Iceberg: Branches of Merge/Update statements should be committed atomically

### PR DESCRIPTION
**Pull request by simhadri-g**
Iceberg: Branches of Merge/Update statements should be committed atomically

### What changes were proposed in this pull request?

Branches of Merge/Update statements should be committed atomically

### Why are the changes needed?

Merge and update queries call commit table twice per query as a result they create 2 snapshots. For any concurrent reads , read data from the incorrect snapshot may result in inconsistent results. 

### Does this PR introduce _any_ user-facing change?

No

### Is the change a dependency upgrade?

No

### How was this patch tested?

Q file and manual test

### CODE REVIEW

File: iceberg/iceberg-handler/src/test/queries/positive/iceberg_atomic_merge_update.q [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+INSERT INTO display (skey, language_id, hierarchy_display) VALUES
+  (1090969, 3, 'f9e59bae9b131de1d8f02d887ee91e20-mergeupdated1-updated1'),
+  (1090969, 3, 'f9e59bae9b131de1d8f02d887ee91e20-mergeupdated1-updated1-insertnew1');
+
+
+-- The number of snapshots in history metadata should increase by 1 after merge query.
+select * from default.display.history;
+
+
+MERGE INTO display USING (
```
Comment (zhangbutao) [inline iceberg/iceberg-handler/src/test/queries/positive/iceberg_atomic_merge_update.q:39]: Just a quick glance: is the qtest related to branch? I have not seen any iceberg branch test.

Comment (simhadri-g) [reply iceberg/iceberg-handler/src/test/queries/positive/iceberg_atomic_merge_update.q:39]: The q test is not related to a branch and tag feature. 

In this context, branch refers to :
https://github.com/apache/hive/blame/master/ql/src/java/org/apache/hadoop/hive/ql/parse/SplitMergeSemanticAnalyzer.java#L45-L49

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java [hunk 2]
```diff
@@ -405,45 +380,63 @@ private Set<FileStatus> listForCommits(JobConf jobConf, String jobLocation) thro
...
+    Table table = null;
+    String branchName = null;
+    boolean isOverWrite = false;
+
+    for (JobContext jobContext : jobContextList) {
+      JobConf conf = jobContext.getJobConf();
+      table = Optional.ofNullable(table).orElse(Catalogs.loadTable(conf, catalogProperties));
+
+      branchName = conf.get(InputFormatConfig.OUTPUT_TABLE_SNAPSHOT_REF);
+      isOverWrite = conf.getBoolean(InputFormatConfig.IS_OVERWRITE, false);
```
Comment (kasakrisz) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:405]: Is it possible that some `jobContext` has `isOverwrite` `true` and some of them is `false`? If the answer is yes then this local variable has the last `jobContext`'s value. Is this expected?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergV2.java [hunk 3]
```diff
@@ -646,6 +652,33 @@ public void testDMLFailsForCopyOnMergeDeleteMode() {
...
+            shell.closeSession();
+          });
+    } catch (Throwable ex) {
+      Throwable cause = Throwables.getRootCause(ex);
+      Assert.assertTrue(cause instanceof ValidationException);
+      Assert.assertTrue(cause.getMessage().startsWith("Found new conflicting delete files"));
+    }
+    List<Object[]> res = shell.executeStatement("SELECT * FROM customers WHERE last_name='Changed'");
+    Assert.assertEquals(5, res.size());
+  }
```
Comment (kasakrisz) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergV2.java:680]: Is this test flaky?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java (hunk 2 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:405]: list of job context is grouped by table name to handle multi-insert statements after the split update/merge. so it should be the same

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergV2.java (hunk 3 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergV2.java:680]: why?

Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergV2.java:680]: work fine on local machine

**Review by kasakrisz — APPROVED** [review]
+1 pending tests

**Pull request merged**